## Pandas 

In [1]:
import pandas as pd  #import the pandas library to our notebook

In [15]:
pd.set_option('display.max_columns', 200)  #This command sets the display columns in a dataframe to the number specified
pd.set_option('display.max_rows', 200)   #This command sets the display rows in a dataframe to the number specified

In [26]:
survey = pd.read_csv("Survey_Resp.csv") #this read_csv method in pandas is reading a csv file from local to a data frame

In [27]:
survey.head(10)

RESP_ID  SUR_SCORE         MASKED_IP          CRT_DT COUNTRY_CODE  \
0  13071656          8  MTQ5,Dk1,TMy,A==  1/18/2018 8:48           FI   
1  13071811         10  MTM0Mjk4,TE0OA==  1/18/2018 8:59           FR   
2  13071889          9      MzY4,DgzOTU=  1/18/2018 9:05           AE   
3  13071909         10  MTM4MzU5MjY2Mg==  1/18/2018 9:06           CZ   
4  13071920         10  MzU3ODcz,Tg1OA==  1/18/2018 9:07           IE   
5  13071988         10  Mjk1,jk1ODE4OQ==  1/18/2018 9:11           HU   
6  13072012          8  MTM2ODUxMTE4,g==  1/18/2018 9:12           GB   
7  13072097          7  MzA1ODE0,zYwOQ==  1/18/2018 9:17           IN   
8  13072253         10  MTg0,DEwMzMx,A==  1/18/2018 9:27           GB   
9  13072647         10  MTI5,zcxMTI0Mg==  1/18/2018 9:50           GB   

          REGION  
0        uusimaa  
1  ile-de-france  
2          dubai  
3          praha  
4            NaN  
5       budapest  
6            NaN  
7  uttar pradesh  
8            NaN  
9            NaN

In [28]:
survey.columns   #will get all the column headers available in the dataset

Index(['RESP_ID', 'SUR_SCORE', 'MASKED_IP', 'CRT_DT', 'COUNTRY_CODE',
       'REGION'],
      dtype='object')

In [29]:
headers = ["RESPONSE_ID","SURVEY_SCORE", "MASKED_IP", "SURVEY_DATE", "COUNTRY_CODE", "REGION"]
survey.columns = headers

In [30]:
survey.columns

Index(['RESPONSE_ID', 'SURVEY_SCORE', 'MASKED_IP', 'SURVEY_DATE',
       'COUNTRY_CODE', 'REGION'],
      dtype='object')

In [32]:
survey.dtypes   #This command will get the datatypes for all the columns in the dataset

RESPONSE_ID      int64
SURVEY_SCORE     int64
MASKED_IP       object
SURVEY_DATE     object
COUNTRY_CODE    object
REGION          object
dtype: object

In [33]:
survey.head(10)

RESPONSE_ID  SURVEY_SCORE         MASKED_IP     SURVEY_DATE COUNTRY_CODE  \
0     13071656             8  MTQ5,Dk1,TMy,A==  1/18/2018 8:48           FI   
1     13071811            10  MTM0Mjk4,TE0OA==  1/18/2018 8:59           FR   
2     13071889             9      MzY4,DgzOTU=  1/18/2018 9:05           AE   
3     13071909            10  MTM4MzU5MjY2Mg==  1/18/2018 9:06           CZ   
4     13071920            10  MzU3ODcz,Tg1OA==  1/18/2018 9:07           IE   
5     13071988            10  Mjk1,jk1ODE4OQ==  1/18/2018 9:11           HU   
6     13072012             8  MTM2ODUxMTE4,g==  1/18/2018 9:12           GB   
7     13072097             7  MzA1ODE0,zYwOQ==  1/18/2018 9:17           IN   
8     13072253            10  MTg0,DEwMzMx,A==  1/18/2018 9:27           GB   
9     13072647            10  MTI5,zcxMTI0Mg==  1/18/2018 9:50           GB   

          REGION  
0        uusimaa  
1  ile-de-france  
2          dubai  
3          praha  
4            NaN  
5       budapest  
6            NaN  
7  uttar pradesh  
8            NaN  
9            NaN

In [34]:
survey.isnull() #Identify the missed values and this method will return the a boolean at cell level based on value exists or not

RESPONSE_ID  SURVEY_SCORE  MASKED_IP  SURVEY_DATE  COUNTRY_CODE  REGION
0           False         False      False        False         False   False
1           False         False      False        False         False   False
2           False         False      False        False         False   False
3           False         False      False        False         False   False
4           False         False      False        False         False    True
...           ...           ...        ...          ...           ...     ...
4783        False         False      False        False         False   False
4784        False         False      False        False         False   False
4785        False         False      False        False         False   False
4786        False         False      False        False          True    True
4787        False         False      False        False         False   False

[4788 rows x 6 columns]

In [36]:
surv_miss_val = survey.isnull() #We are creating anew dataframe based on isnull method acting upon the original dataframe

In [41]:
surv_miss_val.columns.values.tolist()

['RESPONSE_ID',
 'SURVEY_SCORE',
 'MASKED_IP',
 'SURVEY_DATE',
 'COUNTRY_CODE',
 'REGION']

In [50]:
surv_miss_val["REGION"].value_counts()  #This method will return the counts for distinct values existing in the column

False    3215
True     1573
Name: REGION, dtype: int64

##### Now let's count the missing values in each column. "True" represents the missing  values and "False" represents the values exists in Dataset

In [53]:
for col_name in surv_miss_val.columns.values.tolist():  # We are taking each column name into this existing variable
    print("Column Name",col_name)                       # Within the loop we will be print each column name
    print(surv_miss_val[col_name].value_counts())        # And he we are passing each column name to the method to get distinct value counts
    print('------------------------------------------------')    #Just a custom line seperator

Column Name RESPONSE_ID
False    4788
Name: RESPONSE_ID, dtype: int64
------------------------------------------------
Column Name SURVEY_SCORE
False    4788
Name: SURVEY_SCORE, dtype: int64
------------------------------------------------
Column Name MASKED_IP
False    4788
Name: MASKED_IP, dtype: int64
------------------------------------------------
Column Name SURVEY_DATE
False    4788
Name: SURVEY_DATE, dtype: int64
------------------------------------------------
Column Name COUNTRY_CODE
False    3994
True      794
Name: COUNTRY_CODE, dtype: int64
------------------------------------------------
Column Name REGION
False    3215
True     1573
Name: REGION, dtype: int64
------------------------------------------------


#### Survey Response Data Set : Missing data findings
Based on the summary above, out of 4788 rows of data, and 2 columns containing missing data:
<ol>
    <li>"COUNTRY_CODE": 794 missing data</li>
    <li>"REGION": 1573 missing data</li>
</ol>

## Data Analysis

In [55]:
survey.columns

Index(['RESPONSE_ID', 'SURVEY_SCORE', 'MASKED_IP', 'SURVEY_DATE',
       'COUNTRY_CODE', 'REGION'],
      dtype='object')

In [57]:
survey[['SURVEY_SCORE','SURVEY_DATE','COUNTRY_CODE','REGION']].describe(include='all')
# We are tryiong to bring the the descriptive stats in the given dataset based on the columns of our interest 
# and we are using describe method from Pandas Library to achieve that

SURVEY_SCORE     SURVEY_DATE COUNTRY_CODE      REGION
count    4788.000000            4788         3994        3215
unique           NaN            4432          101         365
top              NaN  1/9/2018 17:56           US  california
freq             NaN               6         1468         146
mean        8.119674             NaN          NaN         NaN
std         2.435567             NaN          NaN         NaN
min         0.000000             NaN          NaN         NaN
25%         7.000000             NaN          NaN         NaN
50%         9.000000             NaN          NaN         NaN
75%        10.000000             NaN          NaN         NaN
max        10.000000             NaN          NaN         NaN

In [61]:
survey[['COUNTRY_CODE','RESPONSE_ID','REGION']]

COUNTRY_CODE  RESPONSE_ID         REGION
0              FI     13071656        uusimaa
1              FR     13071811  ile-de-france
2              AE     13071889          dubai
3              CZ     13071909          praha
4              IE     13071920            NaN
...           ...          ...            ...
4783           US     14279849        georgia
4784           US     14280186        florida
4785           US     14280278           ohio
4786          NaN     14281021            NaN
4787           US     14281427          texas

[4788 rows x 3 columns]

In [83]:
a = survey.sort_values(by=['COUNTRY_CODE'],ascending=False, na_position='last')

In [84]:
a.head(10)

RESPONSE_ID  SURVEY_SCORE         MASKED_IP      SURVEY_DATE  \
1214     13010614            10  Mjc2MTk1Mjk0OQ==   1/16/2018 9:39   
1356     13044588             9  Mjc4,Dk3MzUz,w==  1/17/2018 11:56   
4256     14082532             9      ,jk3,TAw,TE4  2/22/2018 11:48   
3273     14337959            10  MTcy,zU4MzI2,w==   3/2/2018 10:40   
2432     13403642            10      ,jk4Mzc5,DM1  1/30/2018 13:13   
3606     13843072             9      ,jg3OTEzOTAw  2/14/2018 13:27   
3593     13837512             9  MTc2MzQ2,zk0MA==   2/14/2018 9:22   
1360     13045031            10  MzMx,zE4MTQyMA==  1/17/2018 12:18   
1679     13693382            10      ,jk2OTMz,zk1   2/9/2018 11:27   
2455     13405934             4      ,jk4,jIwMzc4  1/30/2018 14:37   

     COUNTRY_CODE        REGION  
1214           ZM        lusaka  
1356           ZA           NaN  
4256           ZA           NaN  
3273           ZA           NaN  
2432           ZA           NaN  
3606           ZA           NaN  
3593           ZA           NaN  
1360           ZA           NaN  
1679           ZA           NaN  
2455           ZA  western cape

In [88]:
a.to_csv('surv_sorted.csv')   # The syntax is <dataframe>.to_csv('path/filename.csv')

In [89]:
b=pd.read_csv('surv_sorted.csv')

In [90]:
b.head(10)

Unnamed: 0  RESPONSE_ID  SURVEY_SCORE         MASKED_IP      SURVEY_DATE  \
0        1214     13010614            10  Mjc2MTk1Mjk0OQ==   1/16/2018 9:39   
1        1356     13044588             9  Mjc4,Dk3MzUz,w==  1/17/2018 11:56   
2        4256     14082532             9      ,jk3,TAw,TE4  2/22/2018 11:48   
3        3273     14337959            10  MTcy,zU4MzI2,w==   3/2/2018 10:40   
4        2432     13403642            10      ,jk4Mzc5,DM1  1/30/2018 13:13   
5        3606     13843072             9      ,jg3OTEzOTAw  2/14/2018 13:27   
6        3593     13837512             9  MTc2MzQ2,zk0MA==   2/14/2018 9:22   
7        1360     13045031            10  MzMx,zE4MTQyMA==  1/17/2018 12:18   
8        1679     13693382            10      ,jk2OTMz,zk1   2/9/2018 11:27   
9        2455     13405934             4      ,jk4,jIwMzc4  1/30/2018 14:37   

  COUNTRY_CODE        REGION  
0           ZM        lusaka  
1           ZA           NaN  
2           ZA           NaN  
3           ZA           NaN  
4           ZA           NaN  
5           ZA           NaN  
6           ZA           NaN  
7           ZA           NaN  
8           ZA           NaN  
9           ZA  western cape